### DISCUSSION TITLE
Multiple actions in a keybind

### DISCUSSION THREAD

**Discussion by rmeno12** [ghostty-org/ghostty#3175] [Feature Requests, Ideas] [answered]
Is it possible to bind multiple actions in to a single trigger in a keybind? I didn't see this listed in the docs, but it would be nice to be able to have something like `keybind = ctrl+shift+f=toggle_fullscreen,toggle_window_decorations` to toggle both the fullscreen mode and window decorations at the same time.

**Comment by mitchellh** [comment]
I like this idea a lot. I'll come back to this to think more about the syntax then we can open an issue to track it.

**Comment by citrus-it** [comment]
I'm also interested in this, with my initial use case being something like `shift+cmd+n=new_tab;text:ssh server...`

**Reply by shinebayar-g** [reply]
Have you ever found a way to do this? Even after https://github.com/ghostty-org/ghostty/pull/10015

```ini
keybind = super+shift+t=new_tab
keybind = chain=text:ssh server\n
```

would run ssh in the original tab, not in the new tab.. Tested things like `keybind = chain=next_tab`, `last_tab`, `goto_tab` in between, no luck.

**Reply by kawarimidoll** [reply]
Currently, as noted in this comment https://github.com/ghostty-org/ghostty/issues/9961#issuecomment-3672482332 , key chaining appears to be limited to the same surface only.

I also recently created a discussion about a similar issue.
https://github.com/ghostty-org/ghostty/discussions/10919

**Comment by clarkema** [comment]
I just came here to open exactly the same request, so +1 from me.

Alternatively, since "fullscreen" for me never means "with the decorations visible" maybe the `window-decoration` option could be expanded?

Either `window-decoration` could offer another choice: `true`, `false` and `true-when-windowed`, or alternatively leave `window-decoration` as-is and introduce another option `window-decoration-fullscreen`, also offering `true` and `false` and defaulting to `window-decoration` if unset.

**Reply by alexbet** [reply]
+1

**Reply by rhodes-b** [reply]
> Alternatively, since "fullscreen" for me never means "with the decorations visible" maybe the window-decoration option could be expanded?

If you are on mac below is in the docs
> On macOS, this setting does not work if window-decoration is set to "false", because native fullscreen on macOS requires window decorations to be set.

If you are on linux we have been talking about that here
https://github.com/ghostty-org/ghostty/discussions/4224#discussioncomment-11711228

**Reply by clarkema** [reply]
@rhodes-b  Thanks for the pointer! I'm on Linux/KDE, and having read through the linked thread it turns out that `gtk-titlebar = false` does exactly what I want (at least on ghostty 1.0.0; I'll have to re-check on 1.0.1)

That resolves the fullscreen issue with decorations _and_ means that ghostty windows have the same titlebar (and widgets) as all the rest of my windows, rather than standing out with their GTK titlebar. Perfect!

**Comment by piechologist** [comment]
*Carrying over from discussion #4830 as suggested.*

We have the `performable:` prefix now and I'd like to fire actions conditionally. This could be done by adding the short circuit operators `||` and `&&`. With `;` as action separator, the syntax would look like shell syntax.

For instance:
```sh
keybind = cmd+c=copy_to_clipboard||esc:C
```
If there's a selection copy it to the clipboard, with no selection send the escape sequence to the shell. The `performable:` prefix is not required in this case.

**Reply by zapman449** [reply]
Carrying over from #5325 

in this syntax:

```
keybind = super+f=text:vim &&write_scrollback_file:paste&&text:\n
```

would open the current scrollback history in vim...

**Reply by kaddkaka** [reply]
Similarly, as per my question in discord: https://discord.com/channels/1005603569187160125/1358317722638876863

I would like to create a binding that does
1. grab the text from the last output (copy that or write it to a file)
2. open my editor with that text loaded

**Comment by lfilho** [comment]
+1 as well! My use case is setting a keybind to create a new window or tab and invoking a specific command/alias, so i can make a `global:` keybind open a new window and immediately invoke my note taking setup with vim. So if i'm in a meeting, i can quickly refer to or write new notes

**Comment by amterp** [comment]
+1, my use case was that I was hoping to have a single keybind for "increasing split size", rather than needing 4 for left,right,up,down. Something like

```
keybind = cmd+shift+plus=resize_split:left,50+resize_split:down,50+resize_split:up,50+resize_split:right,50
```

(though I don't know if this would work in practice or if they'd just contradict). Incidentally, if anyone has suggestions on how I can accomplish this another way, I'm all ears!

**Comment by kaddkaka** [comment]
Would scripting or plugin support supersede this idea?

For example creating a function that does two commands after each other and make a keybinding for this?

```
toggle_fullscreen
toggle_window_decorations
```

Does this idea warrant its own existence by supplying a convenient way to express `+`-combined actions? Or will it just be confusing to have a special way to do it when/if there is a more general way to do/script complex actions?

**Comment by atahrijouti** [comment]
I can't think of any way to have ghostty open a split down, launch lazygit, and tie it all up to a keybinding.

I hope sequential actions could solve this for me.

**Reply by rhodes-b** [reply]
#2353 is what you are looking for

this is definitely not the correct discussion to have ghostty run a command for you since that does not exist (this is just doing multiple things in a single keybind)

**Reply by atahrijouti** [reply]
True 👍  partly yes though, but I imagine it's a composite of three features.

- do multiple actions at once (non existing feature)
- first action : split a pane (existing action)
- second action : send characters to existing pane action (not sure exists but i think it doesn't as of yet. This will simply input `lazygit\n`) 

And thus I think maybe I'm still in the right discussion if you see what i mean. Maybe I just got too deep into proposing implementations already 🙈

**Comment by kawarimidoll** [comment]
It would be great if multiple actions could be chained. For now, I'm managing by emulating multiple keys.
https://gist.github.com/kawarimidoll/a8d1d99d182f778c397e23d69ba00685

**Comment by mitchellh** [comment]
@risenowrise Hid your slop. It adds nothing to this conversation.

**Reply by risenowrise** [reply]
fair. i am new to this. eager to learn :)

**Comment by mitchellh** [comment]
I really want to tackle this soon. Every syntax I can think of complicates parsing binding actions which by this point can contain just about anything. So, instead, I thought of something different. Here is some proposed syntax for OP's example:

```
keybind = ctrl+shift+f=unbind
keybind = chain=toggle_fullscreen
keybind = chain=toggle_window_decorations
```

The `chain` key would cause the parsed action to be _appended_ to the most recently bound action. Therefore, starting with `unbind` is a good way to ensure that you're starting a new chain and to establish the key. The chain will inherit the flags of the most recently. bound action (performable, all, global, etc.).

This adds a bit of verbosity but enables a bunch of good use cases for us:

- Doesn't complicate parsing actions, there's no ambiguity, quotations, escapes, etc. needed to support this.
- `chain` makes it pretty clear something special is going on. I thought about using a sigil like `|` or `|>` or something but that makes it pretty special. `chain` is pretty clear, I think.
- Compatible with CLIs without ugly quoting, `--keybind=ctrl+shift+f=unbind --keybind=chain=foo --keybind=chain=bar`
- Easy to make it a configuration error to specify any flags on chains

I know its not _ideal_ and the syntax is a bit odd at first glance, but I think its the best option we've got given the constraints.

**Reply by AnkanMisra** [reply]
With the chain syntax, will each action wait for the previous one to be fully ready before running? 
For example, if I create a new split and then send text to it, will the chain wait until the split is active so the text goes to the correct place?

**Reply by mitchellh** [reply]
That'd be the goal.

**Comment by mitchellh** [comment, marked as answer]
Promoted to an accepted issue #9961
